File: cli/src/commands/status.ts
import { Command } from 'commander'
import chalk from 'chalk'
import { NotifyClient } from '../lib/client.js'

export const statusCommand = new Command('status')
  .description('Check the status of a message')
  .argument('<id>', 'Message ID')
  .option('--server <url>', 'Server URL override')
  .option('--token <token>', 'API token override')
  .action(async (id, opts) => {
    const client = new NotifyClient(opts.server, opts.token)
    const result = await client.getMessage(id)

    if (!result.success) {
      console.error(chalk.red(`✗ Error: ${result.error}`))
      process.exit(1)
    }

    const msg = result.data
    console.log()
    console.log(chalk.bold('Message Details'))
    console.log(chalk.gray('─'.repeat(40)))
    console.log(`  ID:         ${msg.id}`)
    console.log(`  Channel:    ${msg.channelType}`)
    console.log(`  To:         ${msg.toAddress}`)
    console.log(`  Subject:    ${msg.subject || '(none)'}`)
    console.log(`  Status:     ${formatStatus(msg.status)}`)
    console.log(`  Retries:    ${msg.retryCount}/${msg.maxRetries}`)
    if (msg.errorMessage) {
      console.log(`  Error:      ${chalk.red(msg.errorMessage)}`)
    }
    if (msg.sentAt) {
      console.log(`  Sent At:    ${new Date(msg.sentAt).toLocaleString()}`)
    }
    console.log(`  Created:    ${new Date(msg.createdAt).toLocaleString()}`)
    console.log()
  })

function formatStatus(status: string): string {
  switch (status) {
    case 'queued': return chalk.yellow(status)
    case 'sending': return chalk.blue(status)
    case 'sent': return chalk.green(status)
    case 'delivered': return chalk.greenBright(status)
    case 'failed': return chalk.red(status)
    case 'dead': return chalk.bgRed.white(status)
    default: return status
  }
}

File: crates/cli/src/commands/status.rs
use clap::Args;

use crate::client::NotifyClient;
use crate::config::load_config;

#[derive(Args)]
pub struct StatusArgs {
    /// Message ID
    id: String,

    /// Server URL override
    #[arg(long)]
    server: Option<String>,

    /// API token override
    #[arg(long)]
    token: Option<String>,
}

pub async fn run(args: StatusArgs) -> anyhow::Result<()> {
    let config = load_config()?;
    let server = args.server.as_deref().unwrap_or(&config.server);
    let token = args.token.as_deref().unwrap_or(&config.token);

    if token.is_empty() {
        eprintln!("Error: no API token configured. Run 'notifyhub config init' first.");
        std::process::exit(1);
    }

    let client = NotifyClient::new(server, token);
    let resp = client.get_message(&args.id).await?;

    if resp.success {
        if let Some(data) = &resp.data {
            println!("ID:         {}", data.get("id").and_then(|v| v.as_str()).unwrap_or("-"));
            println!("Channel:    {}", data.get("channelType").and_then(|v| v.as_str()).unwrap_or("-"));
            println!("To:         {}", data.get("toAddress").and_then(|v| v.as_str()).unwrap_or("-"));
            println!("Subject:    {}", data.get("subject").and_then(|v| v.as_str()).unwrap_or("-"));
            println!("Status:     {}", data.get("status").and_then(|v| v.as_str()).unwrap_or("-"));
            println!("Retries:    {}/{}",
                data.get("retryCount").and_then(|v| v.as_i64()).unwrap_or(0),
                data.get("maxRetries").and_then(|v| v.as_i64()).unwrap_or(5));
            if let Some(err) = data.get("errorMessage").and_then(|v| v.as_str()) {
                if !err.is_empty() {
                    println!("Error:      {err}");
                }
            }
            if let Some(sent) = data.get("sentAt").and_then(|v| v.as_i64()) {
                println!("Sent at:    {}", format_timestamp(sent));
            }
            println!("Created at: {}", format_timestamp(
                data.get("createdAt").and_then(|v| v.as_i64()).unwrap_or(0)
            ));
        }
    } else {
        let err = resp.error.unwrap_or_else(|| "unknown error".to_string());
        eprintln!("Error: {err}");
        std::process::exit(1);
    }

    Ok(())
}

fn format_timestamp(ts: i64) -> String {
    chrono::DateTime::from_timestamp(ts, 0)
        .map(|dt| dt.format("%Y-%m-%d %H:%M:%S UTC").to_string())
        .unwrap_or_else(|| ts.to_string())
}
